README.md
# adaptive_scaffold

> :warning: This package **REQUIRES** a flutter version >= 1.17 since it depends on Material's NavigationRail.

The adaptive_scaffold allows you to easily create a themable responsive scaffold.


## Getting Started

### 1 - Add to pubspec.yaml
```yml
dependencies:
  adaptive_scaffold: ^0.2.1
```
Don't forget to check on [pub](https://pub.dev/packages/adaptive_scaffold) to see the most recent version

### 2 - Update dependencies
```sh
$ flutter packages get
```

### 3 - Use it
```dart
import "packages:adaptive_scaffold/AdaptiveScaffold.dart";
import 'package:flutter/material.dart';


class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int currentIndex;

  @override
  void initState() {
    super.initState();
    currentIndex = 0;
  }
  
  // Creating the routing destinations. In this example, all with and index > 3 will be set in the drawer and will be using the Navigator to handle routing when the screen size is small or medium. Hence requiring an adaptive widget.
  List<AdaptiveScaffoldDestination> destinations = List.generate(
    5,
    (index) => AdaptiveScaffoldDestination(
      title: "Title $index",
      icon: Icons.ac_unit,
      destination: index > 3
          ? AdaptivePushableWidget(index)
          : Container(
              child: Center(
                child: Text("this is $index"),
              ),
            ),
      inDrawer: index > 3,
    ),
  );

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: myTheme,
      home: AdaptiveScaffold(
        appBar: AppBar(
          title: Text("AdaptiveScaffold title"),
        ),
        currentIndex: currentIndex,
        drawerThemeData: Theme.of(context).copyWith(
          canvasColor: Colors.amber.shade300, // Change the background color of the drawer
        ),
        // Give a custom drawer header
        drawerHeader: MyDrawerHeader(),
        railElevation: RailElevation.Backdrop,
        floatingActionButton: FloatingActionButton(
          backgroundColor: Colors.orange,
          child: Icon(
            Icons.add,
            color: Colors.white,
          ),
          onPressed: () => print("Look, I pressed the FAB!"),
        ),
        destinations: destinations,
        // Add a decoration to the drawer listiles
        activeDrawerItemDecoration: BoxDecoration(
          border: Border(
            right: BorderSide(width: 3.0, color: Colors.red.shade500),
          ),
        ),
        onNavigationIndexChange: (index) {
          setState(() => currentIndex = index);
        },
        body: Container(
          color: Colors.grey.shade200,
          child:
              Center(child: destinations.elementAt(currentIndex).destination),
        ),
      ),
    );
  }
}

class MyDrawerHeader extends DrawerHeader {
  MyDrawerHeader({Key key});

  @override
  DrawerHeader build(BuildContext context) {
    return DrawerHeader(
        margin: EdgeInsets.zero,
        decoration: BoxDecoration(
          color: Colors.grey.shade800,
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Text(
              "Adaptive Scaffold",
              style: TextStyle(
                color: Colors.white,
                fontSize: 22,
              ),
            ),
            Text(
              "In a package",
              style: TextStyle(
                color: Colors.grey.shade300,
                fontSize: 16,
              ),
            ),
          ],
        ));
  }
}
```

## Example & Theming
An example including Theming is available in the example folder.



CHANGELOG.md
## [0.2.1] - 2020-05-07
* Version update on readme

## [0.2.0] - 2020-05-07
* Make adaptive_scaffold completely compatible with 1.7.0
* Add a bottom_navbar_theme until flutter 1.18.0-pre 6 (which contains the Material implementation for BottomNavigationBarTheme) reaches stable

## [0.1.0] - 2020-05-07

* Initial Release.
* Package depends on NavigationRail which requires flutter > 1.7.0
* Package also depends on BottomNavigationBarTheme which requires flutter 1.18.0-pre-6

test/AdaptiveScaffold_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:adaptive_scaffold/adaptive_scaffold.dart';

void main() {
  throw ("No Test for now");
}


example/lib/theme.dart
import 'package:flutter/material.dart';

var myTheme = ThemeData(
  primarySwatch: Colors.red,
  //canvasColor: Colors.amber.shade400,
  hoverColor: Colors.red.shade100,
  focusColor: Colors.blue,
  iconTheme: IconThemeData(
    color: Colors.green.shade600,
  ),
  dividerColor: Colors.grey.shade600,
  dividerTheme: DividerThemeData(
    thickness: 1,
    space: 1, // Width of the divider
    //color: Colors.black,
  ),
  /*
  bottomNavigationBarTheme: BottomNavigationBarThemeData(
      backgroundColor: Colors.grey.shade200,
      showUnselectedLabels: false,
      showSelectedLabels: true,
      selectedItemColor: Colors.red.shade500,
      selectedIconTheme: IconThemeData(
        color: Colors.red.shade400,
        size: 24,
      )),
      */
  navigationRailTheme: NavigationRailThemeData(
    backgroundColor: Colors.red.shade500,
    selectedIconTheme: IconThemeData(
      color: Colors.white,
    ),
    selectedLabelTextStyle: TextStyle(
      color: Colors.white,
    ),
    unselectedIconTheme: IconThemeData(
      color: Colors.red.shade200,
    ),
    unselectedLabelTextStyle: TextStyle(
      color: Colors.red.shade200,
    ),
    labelType: NavigationRailLabelType.all,
  ),
);


lib/adaptive_scaffold.dart
library adaptive_scaffold;

import 'package:flutter/material.dart';

part "bottom_navbar_theme.dart";

/// Describes the elevation of the NavigationRail
enum RailElevation {
  Backdrop,
  Divider,
  Elevated,
}

bool isLargeScreen(BuildContext context) {
  return MediaQuery.of(context).size.width > 960.0;
}

bool isMediumScreen(BuildContext context) {
  return MediaQuery.of(context).size.width > 640.0;
}

/// Describes the destination widgets and their placement in the responsive layout.
///
/// Use this class to configure the [AdaptiveScaffold] widget using:
///
/// * title is the name shown below the navigation icon or on the navigation drawer
/// * icon is [IconData], such as Icons.ac_unit
/// * destination is the [Widget] to be shown
/// * inDrawer is optional and set to false. If true and applicable, this will be set in the [Drawer]
class AdaptiveScaffoldDestination {
  final String title;
  final IconData icon;
  final bool inDrawer;
  final Widget destination;

  const AdaptiveScaffoldDestination({
    @required this.title,
    @required this.icon,
    @required this.destination,
    this.inDrawer = false,
  });
}

/// A scaffold widget that adapts to the current display size, displaying a [Drawer],
/// [NavigationRail], or [BottomNavigationBar]. Navigation destinations are
/// defined in the [destinations] parameter.
///
/// * title is the widget displayed in the [appBar]'s title widget field
/// * body is the shown widget. Note: Only Widgets registered in the [BottomNavigationBar] or the [NavigationRail] can be displayed directly. Others will be displayed using the [Navigator] and [MaterialPageRoute]
/// * railElevation uses the [RailElevation] enum to handle using the [NavigationRail] as a backdrop
/// * onNavigationIndexChange is a callback to be registered to handle navigation from the stateful parent. A sample implementation using an index would looks like:
///
/// ```dart
///   onNavigationIndexChange: (index) {
///      setState(() => currentIndex = index);
///   },
/// ```
///
/// NOTE: This will be removed when flutter 1.18 reaches stable and the BottomNavigationBar can be styled directly from Material's theme.
/// * bottomNavigationThemeData is a [BottomNavBarThemeData] applied to the bottomNavBar.
///   Because of
///
/// * activeDrawerItemDecoration is a [BoxDecoration] applied to the active drawer items. The following example adds a right side blue border to the active item:
///
///```dart
/// ...// Previous code
/// activeDrawerItemDecoration: BoxDecoration(
///   border: Border(
///     right: BorderSide(width: 6.0, color: Colors.blue),
/// ))
///```
///
/// * The drawer can be customized through drawerTheme using a [Theme]. the theme will default the context.
/// Consider using copyWith for easier tweaking.
/// ```dart
/// Theme.of(context).copyWith(
///   // Your theme data
///   canvasColor: Colors.yellow, // Will change the background Color
/// )
/// ```
///
/// General theming for the [NavigationRail]  is done through the Application's [ThemeData]
class AdaptiveScaffold extends StatefulWidget {
  final AppBar appBar;
  final Widget body;
  final RailElevation railElevation;
  final List<AdaptiveScaffoldDestination> destinations;
  final ValueChanged<int> onNavigationIndexChange;
  final FloatingActionButton floatingActionButton;

  final DrawerHeader drawerHeader;
  final ThemeData drawerThemeData;
  final BottomNavBarThemeData bottomNavigationThemeData;

  final int currentIndex;
  final BoxDecoration activeDrawerItemDecoration;

  AdaptiveScaffold({
    this.appBar,
    this.body,
    this.railElevation = RailElevation.Divider,
    this.drawerHeader,
    this.activeDrawerItemDecoration,
    this.bottomNavigationThemeData,
    this.drawerThemeData,
    @required this.currentIndex,
    @required this.destinations,
    this.onNavigationIndexChange,
    this.floatingActionButton,
  });

  @override
  _AdaptiveScaffoldState createState() => _AdaptiveScaffoldState();
}

class _AdaptiveScaffoldState extends State<AdaptiveScaffold> {
  @override
  Widget build(BuildContext context) {
    // Show a Drawer
    if (isLargeScreen(context)) {
      return Row(
        children: [
          buildDrawer(widget.destinations, isLargeDrawer: true),
          widget.railElevation != RailElevation.Backdrop
              ? VerticalDivider()
              : Container(),
          Expanded(
            child: Scaffold(
              appBar: widget.appBar,
              body: widget.body,
              floatingActionButton: widget.floatingActionButton,
              floatingActionButtonLocation: FloatingActionButtonLocation.endTop,
            ),
          ),
        ],
      );
    }

    // Show a navigation rail
    if (isMediumScreen(context)) {
      return Scaffold(
        appBar: widget.appBar,
        drawer: widget.destinations.where((d) => d.inDrawer == true).isNotEmpty
            ? buildDrawer(
                widget.destinations.where((d) => d.inDrawer == true).toList())
            : null,
        body: Row(
          children: [
            NavigationRail(
              leading: widget.floatingActionButton,
              destinations: [
                ...widget.destinations.where((d) => d.inDrawer == false).map(
                      (d) => NavigationRailDestination(
                        icon: Icon(d.icon),
                        label: Text(d.title),
                      ),
                    ),
              ],
              selectedIndex: widget.currentIndex,
              onDestinationSelected: widget.onNavigationIndexChange ?? (_) {},
            ),
            widget.railElevation != RailElevation.Backdrop
                ? VerticalDivider(
                    width: 1,
                    thickness: 1,
                    color: Colors.grey[300],
                  )
                : Container(),
            Expanded(
              child: widget.body,
            ),
          ],
        ),
      );
    }

    // Show a bottom app bar
    return Scaffold(
      body: widget.body,
      drawer: widget.destinations.where((d) => d.inDrawer == true).isNotEmpty
          ? buildDrawer(
              widget.destinations.where((d) => d.inDrawer == true).toList())
          : null,
      appBar: widget.appBar,
      bottomNavigationBar: BottomNavigationBar(
        backgroundColor: widget.bottomNavigationThemeData?.backgroundColor ??
            Theme.of(context).backgroundColor,
        elevation: widget.bottomNavigationThemeData?.elevation ?? 16.0,
        selectedIconTheme:
            widget.bottomNavigationThemeData?.selectedIconTheme ??
                Theme.of(context).iconTheme,
        unselectedIconTheme:
            widget.bottomNavigationThemeData?.unselectedIconTheme ??
                Theme.of(context).iconTheme.copyWith(
                      color: Theme.of(context).unselectedWidgetColor,
                    ),
        selectedItemColor:
            widget.bottomNavigationThemeData?.selectedItemColor ??
                Theme.of(context).primaryColor,
        unselectedItemColor:
            widget.bottomNavigationThemeData?.unselectedItemColor ??
                Theme.of(context).unselectedWidgetColor,
        selectedLabelStyle:
            widget.bottomNavigationThemeData?.selectedLabelStyle ??
                Theme.of(context).textTheme.caption.copyWith(
                      color: Theme.of(context).primaryColor,
                    ),
        unselectedLabelStyle:
            widget.bottomNavigationThemeData?.unselectedLabelStyle ??
                Theme.of(context).textTheme.caption.copyWith(
                      color: Theme.of(context).unselectedWidgetColor,
                    ),
        type: BottomNavigationBarType.fixed,
        showSelectedLabels:
            widget.bottomNavigationThemeData?.showSelectedLabels ?? true,
        showUnselectedLabels:
            widget.bottomNavigationThemeData?.showUnselectedLabels ?? true,
        /*
            BottomNavigationBarTheme.of(context)?.showSelectedLabels ??
                true, // NOTE: Little hack around the base Material to always permit hiding the selected label
                */
        //
        items: [
          ...widget.destinations.where((d) => d.inDrawer == false).map(
                (d) => BottomNavigationBarItem(
                  icon: Icon(
                    d.icon,
                  ),
                  title: Text(
                    d.title,
                  ),
                ),
              ),
        ],
        currentIndex: widget.currentIndex,
        onTap: widget.onNavigationIndexChange,
      ),
      floatingActionButton: widget.floatingActionButton,
    );
  }

  void _destinationTapped(
    AdaptiveScaffoldDestination destination, {
    @required bool needsNavigator,
  }) {
    var idx = widget.destinations.indexOf(destination);
    if (needsNavigator) {
      // NOTE: If not in a LARGE window, you can't navigate to a widget that is not available in the NavigationRail or the BottomNavigationBar
      Navigator.push(context,
          MaterialPageRoute(builder: (content) => destination.destination));
      return;
    }
    if (!isActive(idx)) {
      widget.onNavigationIndexChange(idx);
    }
  }

  bool isActive(int index) {
    return widget.currentIndex == index;
  }

  buildDrawer(
    List<AdaptiveScaffoldDestination> destinations, {
    bool isLargeDrawer = false,
  }) {
    var activeStyle = widget.activeDrawerItemDecoration;
    return Theme(
      data: widget.drawerThemeData ?? Theme.of(context),
      child: Drawer(
        child: Column(
          children: [
            widget?.drawerHeader ?? Container(),
            Expanded(
                child: ListView(
              padding: EdgeInsets.zero,
              children: [
                Container(
                  height: widget?.drawerHeader != null ? 8.0 : 0.0,
                ),
                for (var d in destinations)
                  Container(
                    decoration:
                        widget.destinations.indexOf(d) == widget.currentIndex
                            ? activeStyle
                            : BoxDecoration(),
                    child: ListTile(
                      leading: Icon(d.icon),
                      title: Text(d.title),
                      selected:
                          widget.destinations.indexOf(d) == widget.currentIndex,
                      onTap: () {
                        // Closing the drawer or else it will still be opened when we navigate back
                        if (!isLargeDrawer) {
                          Navigator.of(context).pop();
                        }
                        _destinationTapped(d, needsNavigator: !isLargeDrawer);
                      },
                    ),
                  )
              ],
            ))
          ],
        ),
      ),
    );
  }
}


lib/bottom_navbar_theme.dart
part of adaptive_scaffold;

/// BottomNavBarTheme is a placeholder for the BottomNavigationBar theme until flutter 1.18.0-pre6 becomes mainstream
///
/// /!\ No initial value is given. When used in [AdaptiveScaffold], values that are not set have been matched to the closest meaningful [Theme] property I could think of. See [_AdaptiveScaffoldState]'s small screen builder to see said values.
///
/// TODO: Remove when flutter 1.18 comes around
@immutable
class BottomNavBarThemeData {
  const BottomNavBarThemeData({
    this.backgroundColor,
    this.elevation,
    this.selectedIconTheme,
    this.unselectedIconTheme,
    this.selectedItemColor,
    this.unselectedItemColor,
    this.selectedLabelStyle,
    this.unselectedLabelStyle,
    this.showSelectedLabels,
    this.showUnselectedLabels,
  });

  /// The color of the [BottomNavigationBar] itself.
  ///
  /// See [BottomNavigationBar.backgroundColor].
  final Color backgroundColor;

  /// The z-coordinate of the [BottomNavigationBar].
  ///
  /// See [BottomNavigationBar.elevation].
  final double elevation;

  /// The size, opacity, and color of the icon in the currently selected
  /// [BottomNavigationBarItem.icon].
  ///
  /// See [BottomNavigationBar.selectedIconTheme].
  final IconThemeData selectedIconTheme;

  /// The size, opacity, and color of the icon in the currently unselected
  /// [BottomNavigationBarItem.icon]s.
  ///
  /// See [BottomNavigationBar.unselectedIconTheme].
  final IconThemeData unselectedIconTheme;

  /// The color of the selected [BottomNavigationBarItem.icon] and
  /// [BottomNavigationBarItem.label].
  ///
  /// See [BottomNavigationBar.selectedItemColor].
  final Color selectedItemColor;

  /// The color of the unselected [BottomNavigationBarItem.icon] and
  /// [BottomNavigationBarItem.label]s.
  ///
  /// See [BottomNavigationBar.unselectedItemColor].
  final Color unselectedItemColor;

  /// The [TextStyle] of the [BottomNavigationBarItem] labels when they are
  /// selected.
  ///
  /// See [BottomNavigationBar.selectedLabelStyle].
  final TextStyle selectedLabelStyle;

  /// The [TextStyle] of the [BottomNavigationBarItem] labels when they are not
  /// selected.
  ///
  /// See [BottomNavigationBar.unselectedLabelStyle].
  final TextStyle unselectedLabelStyle;

  /// Whether the labels are shown for the unselected [BottomNavigationBarItem]s.
  ///
  /// See [BottomNavigationBar.showSelectedLabels].
  final bool showSelectedLabels;

  /// Whether the labels are shown for the selected [BottomNavigationBarItem].
  ///
  /// See [BottomNavigationBar.showUnselectedLabels].
  final bool showUnselectedLabels;

  /// Creates a copy of this object but with the given fields replaced with the
  /// new values.
  BottomNavBarThemeData copyWith({
    Color backgroundColor,
    double elevation,
    IconThemeData selectedIconTheme,
    IconThemeData unselectedIconTheme,
    Color selectedItemColor,
    Color unselectedItemColor,
    TextStyle selectedLabelStyle,
    TextStyle unselectedLabelStyle,
    bool showSelectedLabels,
    bool showUnselectedLabels,
  }) {
    return BottomNavBarThemeData(
      backgroundColor: backgroundColor ?? this.backgroundColor,
      elevation: elevation ?? this.elevation,
      selectedIconTheme: selectedIconTheme ?? this.selectedIconTheme,
      unselectedIconTheme: unselectedIconTheme ?? this.unselectedIconTheme,
      selectedItemColor: selectedItemColor ?? this.selectedItemColor,
      unselectedItemColor: unselectedItemColor ?? this.unselectedItemColor,
      selectedLabelStyle: selectedLabelStyle ?? this.selectedLabelStyle,
      unselectedLabelStyle: unselectedLabelStyle ?? this.unselectedLabelStyle,
      showSelectedLabels: showSelectedLabels ?? this.showSelectedLabels,
      showUnselectedLabels: showUnselectedLabels ?? this.showUnselectedLabels,
    );
  }
}


example/lib/PushableRoute.dart
import 'package:adaptive_scaffold/adaptive_scaffold.dart';
import 'package:flutter/material.dart';

class ExamplePushable extends StatelessWidget {
  final int index;
  const ExamplePushable(this.index, {Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    if (isLargeScreen(context)) {
      return PageBody(
        index: index,
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: Text("Hello from pushed route"),
      ),
      body: PageBody(index: index),
    );
  }
}

class PageBody extends StatelessWidget {
  const PageBody({
    Key key,
    @required this.index,
  }) : super(key: key);

  final int index;

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Center(
        child: Text("With index: $index"),
      ),
    );
  }
}


example/lib/main.dart
import 'package:example/PushableRoute.dart';
import 'package:example/theme.dart';
import 'package:flutter/material.dart';
import 'package:adaptive_scaffold/adaptive_scaffold.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  // This widget is the root of your application.
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int currentIndex;

  @override
  void initState() {
    super.initState();
    currentIndex = 0;
  }

  List<AdaptiveScaffoldDestination> destinations = List.generate(
    5,
    (index) => AdaptiveScaffoldDestination(
      title: "Title $index",
      icon: Icons.ac_unit,
      destination: index > 3
          ? ExamplePushable(index)
          : Container(
              child: Center(
                child: Text("this is $index"),
              ),
            ),
      inDrawer: index > 3,
    ),
  );

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: myTheme,
      home: AdaptiveScaffold(
        appBar: AppBar(
          title: Text("AdaptiveScaffold title"),
        ),
        currentIndex: currentIndex,
        drawerHeader: MyDrawerHeader(),
        railElevation: RailElevation.Backdrop,
        //drawerThemeData: ThemeData(),

        floatingActionButton: FloatingActionButton(
          backgroundColor: Colors.orange,
          child: Icon(
            Icons.add,
            color: Colors.white,
          ),
          onPressed: () => print("I pressed the Button"),
        ),
        destinations: destinations,
        activeDrawerItemDecoration: BoxDecoration(
          border: Border(
            right: BorderSide(width: 3.0, color: Colors.red.shade500),
          ),
        ),
        onNavigationIndexChange: (index) {
          setState(() => currentIndex = index);
        },
        body: Container(
          color: Colors.grey.shade200,
          child:
              Center(child: destinations.elementAt(currentIndex).destination),
        ),
      ),
    );
  }
}

class MyDrawerHeader extends DrawerHeader {
  MyDrawerHeader({Key key});

  @override
  DrawerHeader build(BuildContext context) {
    return DrawerHeader(
        margin: EdgeInsets.zero,
        decoration: BoxDecoration(
          color: Colors.grey.shade800,
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Text(
              "Adaptive Scaffold",
              style: TextStyle(
                color: Colors.white,
                fontSize: 22,
              ),
            ),
            Text(
              "In a package",
              style: TextStyle(
                color: Colors.grey.shade300,
                fontSize: 16,
              ),
            ),
          ],
        ));
  }
}
